ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -330,27 +302,69 @@ static VirtualizationRootHandle FindRootAtVnode_Locked(vnode_t vnode, uint32_t v
     return RootHandle_None;
 }
 
-// Returns negative value if it failed, or inserted index on success
-static VirtualizationRootHandle InsertVirtualizationRoot_Locked(PrjFSProviderUserClient* userClient, pid_t clientPID, vnode_t vnode, uint32_t vid, FsidInode persistentIds, const char* path)
+static VirtualizationRootHandle FindOrInsertVirtualizationRoot_LockedMayUnlock(vnode_t virtualizationRootVNode, uint32_t rootVid, FsidInode persistentIds, const char* path)
 {
-    VirtualizationRootHandle rootIndex = FindUnusedIndexOrGrow_Locked();
-    
-    if (RootHandle_None != rootIndex)
+    VirtualizationRootHandle rootIndex;
+    do
     {
-        assert(rootIndex < s_maxVirtualizationRoots);
-        VirtualizationRoot* root = &s_virtualizationRoots[rootIndex];
+        rootIndex = FindRootAtVnode_Locked(virtualizationRootVNode, rootVid, persistentIds);
+        if (rootIndex >= 0)
+        {
+            return rootIndex;
+        }
+        
+        rootIndex = FindUnusedIndex_Locked();
+        if (rootIndex < 0)
+        {
+            // No space, resize array
+            uint16_t newLength = MIN(s_maxVirtualizationRoots * 2u, INT16_MAX + 1u);
+            if (newLength <= s_maxVirtualizationRoots)
+            {
+                KextLog_Error("FindOrInsertVirtualizationRoot_LockedMayUnlock: growing virtualization root array for root at '%s' failed: out of valid indices, array has reached %u items\n", path, s_maxVirtualizationRoots);
+                return RootHandle_None;
+            }
+        
+            VirtualizationRoot* grownArray = Memory_AllocArrayNoBlock<VirtualizationRoot>(newLength);
+            if (grownArray != nullptr)
+            {
+                KextLog_Info("FindOrInsertVirtualizationRoot_LockedMayUnlock: growing virtualization root array for root at '%s' with non-blocking alloc succeeded\n", path);
+                rootIndex = s_maxVirtualizationRoots;
+            }
+            else
+            {
+                // Must drop lock to safely allocate memory with blocking alloc
+                RWLock_ReleaseExclusive(s_virtualizationRootsLock);
```

Comment:
I believe I understand why we need to release the lock, but it seems risky.  Could it cause us to get bad data?

Reply:
I've satisfied myself that this code is not accessing any shared state protected by the lock outside of that lock being held. Beyond that, I've checked that the callers of `FindOrInsertVirtualizationRoot_LockedMayUnlock()` don't make any assumptions about the state before the call being consistent with state after the call - i.e. unlocking temporarily would not cause any inconsistencies. I gave the function the awkward name suffix so that developers (including my forgetful self) are made painfully aware that such assumptions must not be made in future changes.

Have I got it right? I don't know, please do try to poke holes in my code!

Reply:
Incidentally, not having to keep writing/reviewing complicated code like this every time we need to resize an array is why I do still like the generic approach.

Reply:
This all sounds good @pmj thank you!  I should be back full time on this tomorrow and will go thru the comments more.

Reply:
>Have I got it right? I don't know, please do try to poke holes in my code! The main sensitive edge case is where 1 thread decides the data structure needs resizing, and drops the lock to alloc some memory. Meanwhile, another thread jumps on the lock during this time, _also_ decides the array needs to grow, so _also_ drops the lock and allocs some memory. Now we have 2 chunks of memory for growing the array. The condition a few lines down is what's supposed to catch and untangle that situation.

There's only one other scenario that's coming to mind, and it seems extremely unlikely.  But @pmj I want to throw it out there for you to confirm that it's unlikely:

1. The no-block alloc fails
2. We release `s_virtualizationRootsLock` and call the blocking alloc
3. The allocator kicks off paging operations
4. The paging operations result in a write that calls into our kext
5. In handling the write our kext goes down a path that requires a memory allocation
6. The allocator is called ...

Reply:
> * In handling the write our kext goes down a path that requires a memory allocation

This implies that there's an open descriptor to a file in a virtualisation root that was previously unknown to the kext. The only way this can happen is if the kext has only just been loaded.

> * The allocator is called ...

This won't actually lead to deadlock; the system will acknowledge the need to free up more physical memory, and do some more paging, etc. Eventually, something will need to make progress, because not every single open file can be on a previously-undiscovered virtualisation root.